src/mapper/pkg/resolvers/intents_store.go

```diff
@@ -0,0 +1,130 @@
+package resolvers
+
+import (
+	"encoding/json"
+	"github.com/otterize/network-mapper/src/mapper/pkg/graph/model"
+	"github.com/samber/lo"
+	"sort"
+	"time"
+)
+
+type ServiceToDestToTimestamp map[model.OtterizeServiceIdentity]map[model.OtterizeServiceIdentity]time.Time
+
+type DiscoveredIntent struct {
+	Source      model.OtterizeServiceIdentity
+	Destination model.OtterizeServiceIdentity
+	Timestamp   time.Time
+}
+
+type IntentsHolderStore struct {
+	serviceMap ServiceToDestToTimestamp
+}
+
+func NewIntentsHolderStore() IntentsHolderStore {
+	return IntentsHolderStore{
+		serviceMap: make(ServiceToDestToTimestamp),
+	}
+}
+
+func (serviceMap ServiceToDestToTimestamp) MarshalJSON() ([]byte, error) {
+	// OtterizeServiceIdentity cannot be serialized as map key in JSON, because it is represented as a map itself
+	// therefore, we serialize the store as a slice of [Key, Value] "tuples"
+	sourceToIntents := make(map[model.OtterizeServiceIdentity][]DiscoveredIntent)
+	for source, destinations := range serviceMap {
+		var intents []DiscoveredIntent
+		for destination, timestamp := range destinations {
+			intents = append(intents, DiscoveredIntent{
+				Source:      source,
+				Destination: destination,
+				Timestamp:   timestamp,
+			})
+		}
+		sourceToIntents[source] = intents
+	}
+	return json.Marshal(lo.ToPairs(sourceToIntents))
+}
+
+func (serviceMap ServiceToDestToTimestamp) UnmarshalJSON(b []byte) error {
+	var pairs []lo.Entry[model.OtterizeServiceIdentity, []DiscoveredIntent]
+	err := json.Unmarshal(b, &pairs)
+	if err != nil {
+		return err
+	}
+	for _, pair := range pairs {
+		src := pair.Key
+		serviceMap[src] = make(map[model.OtterizeServiceIdentity]time.Time)
+		for _, intent := range pair.Value {
+			serviceMap[src][intent.Destination] = intent.Timestamp
+		}
+	}
+	return nil
+}
+
+func (s *IntentsHolderStore) MarshalJSON() ([]byte, error) {
+	return s.serviceMap.MarshalJSON()
+}
+
+func (s *IntentsHolderStore) UnmarshalJSON(b []byte) error {
+	return json.Unmarshal(b, &s.serviceMap)
+}
+
+func (s *IntentsHolderStore) Update(src model.OtterizeServiceIdentity, dest model.OtterizeServiceIdentity, newTimestamp time.Time) bool {
+	updated := false
```

Comment:
This function has a lot of different cases, but I think it's largely because of the map structure it has. If you use something like this:

```
map[SourceDestPair]time.Time

type SourceDestPair struct {
	Source      model.OtterizeServiceIdentity
	Destination model.OtterizeServiceIdentity
}
```

Then you can just do this:
`s.serviceMap[SourceDestPair{Source: src, Destination: dest}] = newTimestamp`

Also, is `updated` necessary? When is newTimestamp older than what we've already seen? That sounds like it should not happen, and if it does, I think that's indicative of some sort of problem, and we shouldn't just ignore the data. If `updated` isn't necessary, this makes reading this function easier - you don't need to ask yourself "when is data not updated? what does the caller do with this output?", because it is always updated, and the caller never changes their behavior (because there is no return value).

These are questions that pop up because of these different `if` conditions that can occur in this code. As a reader it has me wondering if there's something I'm missing. If they aren't real conditions, it's better to eliminate them completely so that it's easier to understand what's happening in the code.

Reply:
This this point really made or the rest way more simpler, thanks.